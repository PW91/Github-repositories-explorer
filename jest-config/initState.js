export default {
  ui: {
    activeUserCard: null,
    lastSearchQuery: null,
    isSearchEmpty: null,
    isLoading: null
  },
  users: []
}