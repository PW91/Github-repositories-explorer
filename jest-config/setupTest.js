import Enzyme from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import 'regenerator-runtime/runtime.js'
import '~/i18n'

Enzyme.configure({
  adapter: new EnzymeAdapter(),
  disableLifecycleMethods: true
})
