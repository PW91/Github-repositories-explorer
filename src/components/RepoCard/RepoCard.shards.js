import styled from '@emotion/styled'
import { SmallHeading } from '~/components/Typography'
import { IcoStar, IcoLink } from '~/styles/icofont'

export const RepoCard = styled.div`
  padding: .5rem 2rem;
  background-color: ${({ theme }) => theme.colors.background.white};
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
  margin: .5rem 0 .5rem 2rem;
  border-radius: ${({ theme }) => theme.borderRadius.main};
`

export const StarsContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  width: 15%;

  ${SmallHeading} {
    margin-top: .5rem;
  }
`

export const RepoHeader = styled.div`
  display: flex;
  align-items: center;
`

export const RepoMeta = styled.div`
  width: 85%;
`

export const StarIcon = styled(IcoStar)`
  font-size: ${({ theme }) => theme.fontSize.icon.normal};
  margin-left: .5rem;
`

export const LinkIcon = styled(IcoLink)`
  font-size: ${({ theme }) => theme.fontSize.icon.normal};
  margin-left: .5rem;
`
