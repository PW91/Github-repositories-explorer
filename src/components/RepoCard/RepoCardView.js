import React from 'react'
import PropTypes from 'prop-types'
import { ExternalLink } from '~/components/Link'
import {
  RepoCard,
  RepoMeta,
  RepoHeader,
  StarIcon,
  LinkIcon,
  StarsContainer
} from '~/components/RepoCard/RepoCard.shards'
import {
  NormalHeading,
  SmallHeading,
  Paragraph
} from '~/components/Typography'

const RepoCardView = ({
  name,
  stars,
  htmlUrl,
  description
}) => (
  <RepoCard data-test='RepoCardView'>
    <RepoMeta>
      <RepoHeader>
        <NormalHeading>{name}</NormalHeading>
        <ExternalLink
          href={htmlUrl}
          target='_blank'
        >
          <LinkIcon />
        </ExternalLink>
      </RepoHeader>
      <Paragraph>{description}</Paragraph>
    </RepoMeta>
    <StarsContainer>
      <SmallHeading>{stars}</SmallHeading>
      <StarIcon />
    </StarsContainer>
  </RepoCard>
)

RepoCardView.propTypes = {
  name: PropTypes.string,
  stars: PropTypes.number,
  htmlUrl: PropTypes.string,
  description: PropTypes.string
}

export default RepoCardView
