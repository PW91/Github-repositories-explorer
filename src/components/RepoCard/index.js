export { default } from '~/components/RepoCard/RepoCardView'
