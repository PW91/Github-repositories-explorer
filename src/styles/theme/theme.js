import * as colors from '~/styles/theme/colors'
import * as commons from '~/styles/theme/commons'

export default {
  colors,
  ...commons
}
