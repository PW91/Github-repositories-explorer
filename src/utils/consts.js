export const API_URL = 'https://api.github.com'

export const BREAKPOINTS = {
  SM: 480,
  MD: 768,
  LG: 992,
  XL: 1300,
}

export const SEARCH_RESULTS_LIMIT = 5

export const API_ACTIONS = {
  USERS_SEARCH: 'USERS_SEARCH',
  USER_REPOS: 'USER_REPOS'
}

export const STORE_ACTIONS = {
  SET_ACTIVE_USER_CARD: 'SET_ACTIVE_USER_CARD',
  SET_LAST_SEARCH_QUERY: 'SET_LAST_SEARCH_QUERY',
  SET_SEARCH_EMPTY_FLAG: 'SET_SEARCH_EMPTY_FLAG',
  SET_LOADING_FLAG: 'SET_LOADING_FLAG',
  SET_USERS: 'SET_USERS',
  SET_USER_REPOS: 'SET_USER_REPOS'
}
