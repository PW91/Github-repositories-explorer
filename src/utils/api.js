import { format } from 'url'
import { toast } from 'react-toastify'
import { parseJSON } from '~/utils/helpers'
import { API_URL, API_ACTIONS } from '~/utils/consts'

const ajaxer = (apiAction, path, method = 'GET', data = null, query = null) => (
  new Promise((resolve, reject) => {
    const sendForm = data && data.constructor === FormData
    const requestOptions = {
      headers: !sendForm && { 'content-type': 'application/json' },
      method,
      ...!!data && { body: sendForm ? data : JSON.stringify(data) },
    }

    const fullPath = `${API_URL}${path}${query ? format({ query }) : ''}`

    fetch(fullPath, requestOptions)
      .then(res => res.status < 400 ? res.text() : { errorText: res.statusText, status: res.status })
      .then(res => {
        if (res.errorText) {
          const { errorText, status } = res
          toast.error(errorText)
          reject({ errorText, status })
        } else {
          const parsedJson = parseJSON(res)
          const mapped = responseFormatter(parsedJson, apiAction)
          resolve(mapped)
        }
      })
  })
)

export const responseFormatter = (data, apiAction) => {
  switch (apiAction) {
    case API_ACTIONS.USERS_SEARCH:
      return data.items.map(({
        login,
        id,
        avatar_url
      }) => ({
        login,
        id,
        avatarUrl: avatar_url
      }))
    case API_ACTIONS.USER_REPOS:
      return data.map(({
        id,
        owner,
        name,
        html_url,
        stargazers_count,
        description
      }) => ({
        id,
        owner: owner.id,
        name,
        htmlUrl: html_url,
        stars: stargazers_count,
        description
      }))
  }
}

export const manageUsersSearch = {
  get: (query) => ajaxer(API_ACTIONS.USERS_SEARCH, '/search/users', 'GET', null, query)
}

export const manageUserRepos = {
  get: (username) => ajaxer(API_ACTIONS.USER_REPOS, `/users/${username}/repos`)
}
