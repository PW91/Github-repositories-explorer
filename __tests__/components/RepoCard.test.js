import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import serializer from 'jest-emotion'
import { ThemeProvider } from 'emotion-theming'
import { findByTestAtrr } from '>/jest-config/utils'
import theme from '~/styles/theme/theme'
import RepoCard from '~/components/RepoCard'

expect.addSnapshotSerializer(serializer)

describe('RepoCard component', () => {

  it('It renders', () => {
    const wrapper = shallow(<RepoCard />)
    expect(findByTestAtrr(wrapper, 'RepoCardView').exists()).toBeTruthy()
  })

  it('Snapshot', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <RepoCard />
      </ThemeProvider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
