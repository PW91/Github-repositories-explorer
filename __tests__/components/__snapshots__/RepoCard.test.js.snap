// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoCard component Snapshot 1`] = `
.emotion-18 {
  padding: .5rem 2rem;
  background-color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: .5rem 0 .5rem 2rem;
  border-radius: .2rem;
}

.emotion-10 {
  width: 85%;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  color: #002442;
  font-size: 1.6rem;
  margin: 0;
  font-weight: 700;
}

.emotion-4 {
  display: inherit;
}

.emotion-2 {
  font-family: 'icofont' !important;
  speak: none;
  font-style: normal;
  font-weight: 400;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: 1.8rem;
  margin-left: .5rem;
}

.emotion-2:before {
  content: "\\f103";
}

.emotion-8 {
  color: #002442;
  font-size: 1.4rem;
  margin: 0;
}

.emotion-16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 15%;
}

.emotion-16 .emotion-13 {
  margin-top: .5rem;
}

.emotion-12 {
  color: #002442;
  font-size: 1.2rem;
  margin: 0;
  font-weight: 700;
}

.emotion-14 {
  font-family: 'icofont' !important;
  speak: none;
  font-style: normal;
  font-weight: 400;
  font-variant: normal;
  text-transform: none;
  line-height: 1;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-size: 1.8rem;
  margin-left: .5rem;
}

.emotion-14:before {
  content: "\\f104";
}

<div
  className="emotion-18 emotion-19"
  data-test="RepoCardView"
>
  <div
    className="emotion-10 emotion-11"
  >
    <div
      className="emotion-6 emotion-7"
    >
      <h3
        className="emotion-0 emotion-1"
      />
      <a
        className="emotion-4 emotion-5"
        target="_blank"
      >
        <span
          className="emotion-2 emotion-3"
        />
      </a>
    </div>
    <p
      className="emotion-8 emotion-9"
    />
  </div>
  <div
    className="emotion-16 emotion-17"
  >
    <h5
      className="emotion-12 emotion-13"
    />
    <span
      className="emotion-14 emotion-15"
    />
  </div>
</div>
`;
